### PR TITLE
Add Cancellation token for file system enumeration

### LINKED ISSUE #299: Use cancellation with Explorer's DirectorySearch
The all files and folders search with Explorer's DirectorySearch seems to run quite long, can we add the use of cancellation token to stop when next query comes in.

### PR DESCRIPTION
**Pull request by taooceros**
fix #299 
Also using EnumerationOption to avoid one access denied location breaking the whole searching similar to #209 .

Outcome:
- Improved Explorer's handling of long running results 
- Enhanced result list update speed

### CODE REVIEW

**Comment by jjw24** [comment]
Ok so found the bottleneck that causes query to run very long for explorer:

image

when you have a lot of results like below, and it's trying to add new results (called from updateresultview) in after using except to exclude the same results currently display, this operation is very costly:
image

**Comment by taooceros** [comment]
That seems what #195 is trying to do. You can take a look on the same method in that pr.

**Comment by jjw24** [comment]
Ok cool, yeah I haven't looked at the code yet, just being taking it for test drives, will do so very soon

**Comment by taooceros** [comment]
I think enumerate all file including subdirectory for about 200000 results may take about 5s and 200MB memory, which is still costly. This should still be added with an cancellation token.

**Comment by taooceros** [comment]
> Ok so found the bottleneck that causes query to run very long from explorer:
> 
> image
> 
> when you have a lot of results like below, and it's trying to add new results (called from updateresultview) in after using except to exclude the same results currently display, this operation is very costly:
> image
> 
> if you let it just display the new results, blazing fast

Also, due to the lock strategy, by doing huge calculation within NewResutls method will hold the lock, which blocks ongoing results view update.

**Comment by jjw24** [comment]
agree and agree, wonder if there is a better way to do this. also maybe if we want to do it in separate pr so cancellation is not blocked by this other fix

**Comment by taooceros** [comment]
> agree and agree, wonder if there is a better way to do this. also maybe if we want to do it in separate pr so cancellation is not blocked by this other fix

That's already exist in #195, so no worry!

**Comment by taooceros** [comment]
I also fix some code that is used for testing in #195, which isn't a bit deal.

**Comment by taooceros** [comment]
After merging the dev, the recursive search for 180000 result takes about 5s to finish, which is significantly faster than before.

**Comment by taooceros** [comment]
From testing, seems that the large amount of result won't be collected by GC even after a few other queries. By changing the List capacity, it should show a memory lower down but seems no effect....
Not sure whether my testing is wrong.

**Comment by taooceros** [comment]
Unintentionally add a test code in PublicAPIInstance due to #300, should remove it before merging.

**Comment by jjw24** [comment]
> Unintentionally add a test code in PublicAPIInstance due to #300, should remove it before merging.

this one?
```
  Application.Current.Dispatcher.Invoke(() =>
            {
                useMainWindowAsOwner = false; <=============
                var msg = useMainWindowAsOwner ? new Msg { Owner = Application.Current.MainWindow } : new Msg();
                msg.Show(title, subTitle, iconPath);
            });
```

**Comment by taooceros** [comment]
> > Unintentionally add a test code in PublicAPIInstance due to #300, should remove it before merging.
> 
> this one?
> 
> ```
>   Application.Current.Dispatcher.Invoke(() =>
>             {
>                 useMainWindowAsOwner = false; <=============
>                 var msg = useMainWindowAsOwner ? new Msg { Owner = Application.Current.MainWindow } : new Msg();
>                 msg.Show(title, subTitle, iconPath);
>             });
> ```

Yes

**Comment by jjw24** [comment]
getting this error ,

repro:
set explorer hotkey to 'e', type 'e a'. first couple of characters gets this error 

image

**Comment by jjw24** [comment]
open flow's settings get this:
image

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs
Comment (jjw24) [inline Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs:80]: in your previous changes i have seen you use ThrowIfCancellationRequested() and catch it with an OperationCancelledException. Is one way better than the other? Should we use IsCancellationRequested way to handle it more elegantly?

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs:80]: Agree with you. I will change that to throw, which can reduce some code for checking null.

**Comment by taooceros** [comment]
> open flow's settings get this:
> image

Forget to add null check😅, will do now.

**Comment by taooceros** [comment]
> getting this error ,
> repro: set explorer hotkey to 'e', type 'e a'. first couple of characters gets this error

I cannot reproduce the error....Could you please provide the stacktrace for that error?

**Comment by jjw24** [comment]
> > getting this error ,
> > repro: set explorer hotkey to 'e', type 'e a'. first couple of characters gets this error
> 
> I cannot reproduce the error....Could you please provide the stacktrace for that error?

This maybe related to the null check, will check again later

**Comment by taooceros** [comment]
@jjw24 Should we try to fix #308 here or use a new pr to fix it?

**Comment by jjw24** [comment]
Separate pr please

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs]: Well I thought iscancellationrequested is nicer and instead of returning null you can just return empty results? Rather than having an extra catch.

Is there a reason for returning null rather than empty result list?

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs]: Just a few difference, not a lot.
We will just throw the error, which will be catched by PluginManager.QueryForPluginAsync, so no worry for doing null check.

**Comment by taooceros** [comment]
> Separate pr please

Sure

File: Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs
Comment (jjw24) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs]: Yeah but cancellation of previous query is a mechanism we require throughout, so probably not great to be treating it like the program errored out. Maybe change this next time in separate pr, for now just leave it as cancellationrequested

Comment (taooceros) [reply Plugins/Flow.Launcher.Plugin.Explorer/Search/DirectoryInfo/DirectoryInfoSearch.cs]: Well actually, throw an taskcancelledexcpetion is what Microsoft suggested in cooperative cancellation. This is similar to how we shall handle operationcancelledexception, which is the base of taskcancelledexcpetion.

However, it seems that in Debug mode, throwing a lot of exception may cause stunt……which is not a bit due, but somehow not promising.

**Comment by jjw24** [comment]
> getting this error ,
> 
> repro:
> set explorer hotkey to 'e', type 'e a'. first couple of characters gets this error
> 
> image

yeah ok still getting this error. Set search activation hotkey to 'e' and just type 'e' i got it.

**Comment by jjw24** [comment]
sorted in https://github.com/Flow-Launcher/Flow.Launcher/pull/301/commits/ff5e3695e9e538ceafc997b96adf2a4f7cb0a4f2

**Comment by taooceros** [comment]
> sorted in https://github.com/Flow-Launcher/Flow.Launcher/pull/301/commits/ff5e3695e9e538ceafc997b96adf2a4f7cb0a4f2

Nice! I understand that why I am unable to reproduce, because I have a few quick access.

**Comment by jjw24** [comment]
yeah haha, i should have said remove roaming profile

**Comment by taooceros** [comment]
> yeah haha, i should have said remove roaming profile

Hahaha 🤣

**Comment by taooceros** [comment]
https://github.com/Flow-Launcher/Flow.Launcher/pull/301/commits/9914124d200ef5988c0488f0f65a9e4bc3082d7e remove the extra checking to make the code clear. We don't need to check whether the quick folder list is empty and don't need to check the action keyword as what we have talked about in Window Walker plugin.

**Comment by taooceros** [comment]
@jjw24 Any further change required?

File: Flow.Launcher/ViewModel/ResultsViewModel.cs
Comment (jjw24) [inline Flow.Launcher/ViewModel/ResultsViewModel.cs:147]: @taooceros do you know if this comment is still relevant?

Comment (taooceros) [reply Flow.Launcher/ViewModel/ResultsViewModel.cs:147]: I think yes, because only results from main result set will be added via resultUpdateQueue, the one for context menu and history is still added directly.
However, I am not sure the deadlock meaning here. Maybe just avoiding potential stunt of main ui due to lock machanism.

Comment (jjw24) [reply Flow.Launcher/ViewModel/ResultsViewModel.cs:147]: cool, thank you

**Review by jjw24 — APPROVED** [review]

**Pull request merged**
